python_modules/dagster/dagster/_core/definitions/asset_graph_subset.py

```diff
@@ -1,70 +1,129 @@
 import operator
 from collections import defaultdict
 from datetime import datetime
-from typing import AbstractSet, Any, Callable, Dict, Iterable, Mapping, Optional, Set, Union, cast
+from functools import cached_property
+from typing import (
+    AbstractSet,
+    Any,
+    Callable,
+    Dict,
+    Iterable,
+    Mapping,
+    NamedTuple,
+    Optional,
+    Set,
+    Union,
+    cast,
+)
 
 from dagster import _check as check
 from dagster._core.definitions.partition import (
     PartitionsDefinition,
     PartitionsSubset,
 )
+from dagster._core.definitions.time_window_partitions import PartitionKeysTimeWindowPartitionsSubset
 from dagster._core.errors import (
     DagsterDefinitionChangedDeserializationError,
 )
 from dagster._core.instance import DynamicPartitionsStore
+from dagster._serdes.serdes import NamedTupleSerializer, whitelist_for_serdes
 
 from .asset_graph import AssetGraph
 from .events import AssetKey, AssetKeyPartitionKey
 
 
-class AssetGraphSubset:
-    def __init__(
-        self,
-        asset_graph: AssetGraph,
-        partitions_subsets_by_asset_key: Optional[Mapping[AssetKey, PartitionsSubset]] = None,
+class AssetGraphSubsetSerializer(NamedTupleSerializer):
+    def before_pack(self, value: "AssetGraphSubset") -> "AssetGraphSubset":
+        converted_partitions_subsets_by_serialized_asset_key = {}
+        for k, v in value.partitions_subsets_by_serialized_asset_key.items():
+            if isinstance(v, PartitionKeysTimeWindowPartitionsSubset):
+                converted_partitions_subsets_by_serialized_asset_key[
+                    k
+                ] = v.to_time_window_partitions_subset()
+            else:
+                converted_partitions_subsets_by_serialized_asset_key[k] = v
+
+        return value._replace(
+            partitions_subsets_by_serialized_asset_key=converted_partitions_subsets_by_serialized_asset_key
+        )
+
+
+@whitelist_for_serdes(serializer=AssetGraphSubsetSerializer)
+class AssetGraphSubset(
+    NamedTuple(
+        "_AssetGraphSubset",
+        [
+            ("partitions_subsets_by_serialized_asset_key", Mapping[str, PartitionsSubset]),
```

Comment:
Unfortunately json cannot serialize dictionaries that aren't keyed by a primitive or a string, so we key by the serialized asset key instead

Reply:
Could it make sense to handle that in the custom serializer so it doesn't need to leak into the data model for the class?

Reply:
Also, don't want to increase scope too much, but it could make sense to try to address this at the serialization layer if it's not too big a lift.

Reply:
> Also, don't want to increase scope too much, but it could make sense to try to address this at the serialization layer if it's not too big a lift.

I did take a stab at implementing it this way.
- The properly serialized asset key `pack_value(asset_key....)` value is a dictionary, which cannot be used as a key to a dictionary
- We could add custom logic in the serialization layer to convert an asset key to a string (i.e. `asset_key.to_user_string()` but this seems like a pain to deal with if we did decide to add additional fields to the AssetKey class

Reply:
>  Could it make sense to handle that in the custom serializer so it doesn't need to leak into the data model for the class?

My initial version actually used this approach, it looked like this: https://github.com/dagster-io/dagster/blob/79652f967fe0bf0c0d7caa68069d5fd53afbb2f0/python_modules/dagster/dagster/_core/definitions/asset_graph_subset.py#L42-L69

I moved away from this implementation because it felt like a duplicate of the existing logic, with the exception of converting the asset key to a string, so maybe it would be better to reduce code surface area by just converting the asset keys to serialized form.

I feel mixed on this though -- I can see either making sense. It certainly is cleaner to not have to convert to/from serialized asset keys.

Reply:
> but this seems like a pain to deal with if we did decide to add additional fields to the AssetKey class

I think that we can be confident that we are not going to do this. I suspect a lot of other places would break as well.